### PR TITLE
[LAT-33] Allowing array of lines for RE pattern

### PR DESCRIPTION
**Pull request by pedrohr**
https://lattice.atlassian.net/browse/LAT-33

### CODE REVIEW

File: src/snovault/schema_utils.py [hunk 1]
```diff
@@ -281,6 +282,19 @@ def notSubmittable(validator, linkTo, instance, schema):
...
+        patrn = ''.join(patrn)
+
+    if (
+        validator.is_type(instance, "string") and
+        not re.search(patrn, instance)
+    ):
+        yield ValidationError("%r does not match %r" % (instance, patrn))
+
+
+
```
Comment (keenangraham) [inline src/snovault/schema_utils.py]: Extra line.

File: src/snovault/schema_utils.py [hunk 2]
```diff
@@ -291,6 +305,7 @@ class SchemaValidator(Draft4Validator):
     VALIDATORS['permission'] = permission
     VALIDATORS['requestMethod'] = requestMethod
     VALIDATORS['uniqueItems'] = uniqueItems
+    VALIDATORS['pattern'] = pattern
```
Comment (keenangraham) [inline src/snovault/schema_utils.py:307]: Kk, overriding the default pattern definition.

File: src/snovault/schema_utils.py [hunk 3]
```diff
@@ -281,6 +282,19 @@ def notSubmittable(validator, linkTo, instance, schema):
...
+# modifying https://github.com/lrowe/jsonschema_serialize_fork/blob/master/jsonschema_serialize_fork/_validators.py#L143
+def pattern(validator, patrn, instance, schema):
+    if isinstance(patrn, list):
+        patrn = ''.join(patrn)
+
+    if (
+        validator.is_type(instance, "string") and
+        not re.search(patrn, instance)
+    ):
+        yield ValidationError("%r does not match %r" % (instance, patrn))
```
Comment (keenangraham) [inline src/snovault/schema_utils.py:294]: Should definitely add unit/integration test for this.

Probably easiest thing to do would be to modify or add your own fake testing schema here: 
https://github.com/Lattice-Data/snovault/blob/d6bec11aa60244172d52512d918fc65663071073/src/snovault/tests/testing_views.py#L277

And then try to post to the endpoint you made with different objects that either pass or fail pattern validation, like here but without the fixture: https://github.com/Lattice-Data/snovault/blob/d6bec11aa60244172d52512d918fc65663071073/src/snovault/tests/test_post_put_patch.py#L78

Or follow this approach and call validate explicitly: https://github.com/Lattice-Data/snovault/blob/d6bec11aa60244172d52512d918fc65663071073/src/snovault/tests/test_schema_utils.py#L17

Can see more examples of testing schema validation on real encoded file objects: https://github.com/ENCODE-DCC/encoded/blob/39478a6bff4dbc1c65c1fd889ffaa0ff0d74d557/src/encoded/tests/test_schema_file.py#L16

**Review by keenangraham — CHANGES_REQUESTED** [review]

File: src/snovault/schema_utils.py [hunk 4]
```diff
@@ -281,6 +282,19 @@ def notSubmittable(validator, linkTo, instance, schema):
     yield ValidationError('submission disallowed')
 
 
+# modifying https://github.com/lrowe/jsonschema_serialize_fork/blob/master/jsonschema_serialize_fork/_validators.py#L143
+def pattern(validator, patrn, instance, schema):
+    if isinstance(patrn, list):
```
Comment (keenangraham) [inline src/snovault/schema_utils.py:287]: It looks like there is dependency in JS form that expects this to be string? https://github.com/Lattice-Data/snovault/blob/d6bec11aa60244172d52512d918fc65663071073/src/snowflakes/static/components/form.js#L293

**Review by keenangraham — CHANGES_REQUESTED** [review]

File: src/snovault/tests/test_schema_utils.py [hunk 5]
```diff
@@ -32,3 +32,39 @@ def test_uniqueItems_validates_normalized_links(content, threadlocals):
...
+    uuid = 'option2'
+    data = [
+        uuid,
+        '/testing-link-targets/{}'.format(uuid),
+    ]
+
+    validated, errors = validate(schema, data)
+    assert len(errors) == 0
+
+    schema['items']['pattern'] = ['option1|', 'option2']
```
Comment (keenangraham) [inline src/snovault/tests/test_schema_utils.py:53]: You probably already thought about just joining on `|` to clean up formatting a bit.

File: src/snovault/schema_utils.py [hunk 6]
```diff
@@ -389,7 +389,8 @@ def combine_schemas(a, b):
         if a[name] == b[name]:
             combined[name] = a[name]
         elif name == 'type':
-            combined[name] = sorted(set(ensurelist(a[name]) + ensurelist(b[name])))
+            combined[name] = sorted(
```
Comment (keenangraham) [inline src/snovault/schema_utils.py:392]: I know this is autoformatting. I would probably format it like: 
```
combined[name] = sorted(
    set(ensurelist(a[name]) + ensurelist(b[name]))
)
```
Not a big deal.

**Review by keenangraham — APPROVED** [review]
Looks good, thanks for adding tests.

File: src/snovault/tests/test_schema_utils.py (hunk 5 above)
Comment (pedrohr) [reply src/snovault/tests/test_schema_utils.py:53]: Yes, that's actually the use case we have. But after options, the last line in our use case is an ID logic for a suffix. We can't join by '|' 😢

File: src/snovault/schema_utils.py (hunk 6 above)
Comment (pedrohr) [reply src/snovault/schema_utils.py:392]: I agree, looks better! Just changed

**Pull request merged**